cli/src/commands/parallelize.rs

```diff
@@ -0,0 +1,208 @@
+        // commits they will appear in the same relative order in `jj log` that
+        // they were in before being parallelized. The new order is children
+        // before parents.
+        .rev()
+        .collect_vec();
+    rebase_descendants(
+        &mut tx,
+        command.settings(),
+        &new_parents,
+        &target_commits_to_rebase,
+        Default::default(),
+    )?;
+
+    tx.finish(
+        ui,
+        format!("parallelized {} commits", target_commit_ids.len()),
+    )
+}
+
+/// Returns the head of the target revset or an error if the revset has multiple
+/// heads.
+fn verify_target_has_a_single_head(
+    target_revset: &Rc<RevsetExpression>,
+    repo: &dyn Repo,
+) -> Result<CommitId, CommandError> {
+    let mut target_heads: Vec<CommitId> = target_revset
+        .heads()
+        .evaluate_programmatic(repo)?
+        .iter()
+        .collect();
+    if target_heads.len() != 1 {
+        let heads = target_heads.iter().map(short_commit_hash).join(", ");
+        return Err(user_error(format!(
+            "Cannot parallelize a set of revisions with multiple heads. Heads: {heads}"
+        )));
+    }
+    Ok(target_heads.pop().unwrap())
+}
+
+/// Returns the root of the target revset or an error if the revset has multiple
+/// roots.
+fn verify_target_has_a_single_root(
+    target_revset: &Rc<RevsetExpression>,
+    repo: &dyn Repo,
+) -> Result<CommitId, CommandError> {
+    let mut target_roots: Vec<CommitId> = target_revset
+        .roots()
+        .evaluate_programmatic(repo)?
+        .iter()
+        .collect();
+    if target_roots.len() != 1 {
+        let roots = target_roots.iter().map(short_commit_hash).join(", ");
+        return Err(user_error(format!(
+            "Cannot parallelize a set of revisions with multiple roots. Roots: {roots}"
+        )));
+    }
+    Ok(target_roots.pop().unwrap())
+}
+
+/// The target commits must be connected. The parents of every commit except the
+/// root commit must also be target commits.
+fn verify_target_commits_are_connected(
+    target_root: &CommitId,
+    target_commits: &[Commit],
+    target_commit_ids: &IndexSet<CommitId>,
+) -> Result<(), CommandError> {
+    for target_commit in target_commits.iter() {
+        if target_commit.id() == target_root {
+            continue;
+        }
+        for parent in target_commit.parent_ids() {
+            if !target_commit_ids.contains(parent) {
+                if target_commit.parent_ids().len() == 1 {
+                    return Err(user_error(
+                        "Cannot parallelize since the target revisions are not connected.",
+                    ));
+                }
+                return Err(user_error(format!(
+                    "Only the root commit is allowed to have parents which are not being \
+                     parallelized.\nCommit {} violates this rule.",
+                    short_commit_hash(target_commit.id())
+                )));
+            }
+        }
+    }
```

Comment:
Maybe better to check if the range is linear connected graph?

With the current simplified `common_parents` lookup, non-linear graph will be parallelize as follows:
```rust
    test_env.jj_cmd_ok(&workspace_path, &["new", "-m=0", "root()"]);
    test_env.jj_cmd_ok(&workspace_path, &["new", "-m=1", "description(0)"]);
    test_env.jj_cmd_ok(&workspace_path, &["new", "-m=2", "description(0)"]);
    test_env.jj_cmd_ok(&workspace_path, &["new", "-m=3", "description(0)"]);
    test_env.jj_cmd_ok(&workspace_path, &["new", "-m=4", "all:heads(..)"]);
    test_env.jj_cmd_ok(&workspace_path, &["new", "-m=1c", "description(1)"]);
    test_env.jj_cmd_ok(&workspace_path, &["new", "-m=2c", "description(2)"]);
    test_env.jj_cmd_ok(&workspace_path, &["new", "-m=3c", "description(3)"]);
    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
    @  b043eb81416c 3c parents: 3
    │ ◉    48277ee9afe0 4 parents: 3 2 1
    ╭─┼─╮
    ◉ │ │  944922f0c69f 3 parents: 0
    │ │ │ ◉  9d28e8e38435 2c parents: 2
    │ ├───╯
    │ ◉ │  97d7522f40e8 2 parents: 0
    ├─╯ │
    │ ◉ │  6c82c22a5e35 1c parents: 1
    │ ├─╯
    │ ◉  0c058af014a6 1 parents: 0
    ├─╯
    ◉  745bea8029c1 0 parents:
    ◉  000000000000 parents:
    "###);

    let (_stdout, stderr) = test_env.jj_cmd_ok(&workspace_path, &["parallelize", "description(0)::description(4)"]);
    insta::assert_snapshot!(stderr, @r###"
    Working copy now at: yostqsxw bec35f45 (empty) 3c
    Parent commit      : rlvkpnrz cbb4e169 (empty) 0
    Parent commit      : kkmpptxz 82918d78 (empty) 1
    Parent commit      : zsuskuln 8f4b8ef6 (empty) 2
    Parent commit      : mzvwutvl cb944786 (empty) 3
    "###);
    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
    ◉      fbc7f2274ecd 2c parents: 0 1 2
    ├─┬─╮
    │ │ │ @  bec35f45fcb1 3c parents: 0 1 2 3
    ╭─┬─┬─┤
    │ │ │ ◉  cb9447869bf0 3 parents:
    │ │ ◉ │  8f4b8ef68676 2 parents:
    │ │ ├─╯
    │ │ │ ◉  1985e0427139 1c parents: 0 1
    ╭─┬───╯
    │ ◉ │  82918d78c984 1 parents:
    │ ├─╯
    ◉ │  cbb4e1692ef4 0 parents:
    ├─╯
    │ ◉  14ca4df576b3 4 parents:
    ├─╯
    ◉  000000000000 parents:
    "###);
```

Reply:
Which part of that do you consider unexpected? It looks good to me as far as I can tell, but I might be missing something.

Reply:
For example, ancestors of `3c` were originally `3c -> 3 -> 0 -> root()`. After parallelized, they become `3c -> (0, 1, 2, 3) -> ..`. I expect they should be either `3c -> (0, 3) -> root()` (preserving ancestry) or `3c -> 3 -> root()` (ignoring ancestry within the range).

Reply:
I see `jj parallelize` as saying "actually, these commits don't depend on each other", but dependencies outside of the set should be preserved. Since `3c` depended on all of `0::3` before, it should still have those as ancestors, even if they no longer internally depend on each other.

Reply:
3c only depended on 3 which only depended on 0. So it shouldn't have 2 as a dependency afterwards, right?

I think Yuya is right that this is broken, and I think it's because the updated common_parents logic depends on an assumption that doesn't hold.

Rather than change the rules for the target set, I'd like to see if we can fix the common parents computation. I'll take a look.

Reply:
> 3c only depended on 3 which only depended on 0. So it shouldn't have 2 as a dependency afterwards, right?

Oh, sorry, I missed that. Yes, I agree that it's broken then.

Reply:
Ok, I fixed this case and added Yuya's test.